### PR TITLE
dicrease the complexity of  CalcDep from exponential to linear

### PR DESCRIPTION
**Pull request by yaochengji**
The DCE pass [bug](https://discuss.tvm.ai/t/problem-in-anf-and-dce-pass/4189) is not caused by infinite loop, it is just time consuming.

@MarisaKirisame  please review, thanks!

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
LGTM.
note that you *should not* pass in a graph to DCE. We only intend for it to support tree form. In general, all the high level analysis/pass (pointer analysis, partial eval, dead code elimination, ad) all assume the input is in tree form.

**Comment by yaochengji** [comment]
Why is it assumed a tree form? I can see that the base class `ExprVisitor` also use the `visit_counter_` to make sure all the expr are visited once.

**Comment by MarisaKirisame** [comment]
@yaochengji because what does it mean for a graph form program to have scope and effect?

**Comment by yaochengji** [comment]
@MarisaKirisame but I think since using a `visit_counter` is a default behavior in `ExprVisitor`, it should not be dropped in `CalcDep `.

cc @icemelon9

**Comment by MarisaKirisame** [comment]
you can still has it. I just mean you should not pass graph in.

**Comment by yaochengji** [comment]
@MarisaKirisame  do you think this commit should be merged?

**Comment by MarisaKirisame** [comment]
sure. It will increase the robustness of the DCE.

**Comment by vinx13** [comment]
Please add a test case

**Comment by yaochengji** [comment]
@vinx13 Done.

File: src/relay/pass/dead_code.cc [hunk 1]
```diff
@@ -110,7 +110,11 @@ class CalcDep : private ExprVisitor {
   VarMap<size_t> use_map_;
 
   void VisitExpr(const Expr& e) final {
-    return ExprFunctor<void(const Expr& e)>::VisitExpr(e);
+    visit_counter_[e.get()]++;
+    if (visit_counter_[e.get()] <= 2) {
```
Comment (vinx13) [inline src/relay/pass/dead_code.cc]: why `<=2`?

Comment (MarisaKirisame) [reply src/relay/pass/dead_code.cc]: the dce code seprate variable into three parts:
used 0 times (remove)
used 1 times (inline)
used 2 times (dont do anything).

Comment (vinx13) [reply src/relay/pass/dead_code.cc]: okay, let's add a comment

Comment (yaochengji) [reply src/relay/pass/dead_code.cc]: Done.

**Comment by vinx13** [comment]
Thanks @yaochengji  @MarisaKirisame this is merged

**Pull request merged**